src/librustc_mir/borrow_check/nll/region_infer.rs

```diff
@@ -33,15 +35,25 @@ pub struct RegionInferenceContext<'tcx> {
     /// regions, these start out empty and steadily grow, though for
     /// each free region R they start out containing the entire CFG
     /// and `end(R)`.
-    liveness_constraints: IndexVec<RegionVid, Region>,
+    ///
+    /// In this `BitMatrix` representation, the rows are the region
+    /// variables and the columns are the free regions and MIR locations.
+    liveness_constraints: BitMatrix,
 
     /// The final inferred values of the inference variables; `None`
     /// until `solve` is invoked.
-    inferred_values: Option<IndexVec<RegionVid, Region>>,
+    inferred_values: Option<BitMatrix>,
 
     /// The constraints we have accumulated and used during solving.
     constraints: Vec<Constraint>,
 
+    /// A map from each MIR Location to its column index in
+    /// `liveness_constraints`/`inferred_values`. (The first N columns are
+    /// the free regions.)
+    point_indices: BTreeMap<Location, usize>,
```

Comment:
... hmm, maybe at some point we'll want an `IndexBitMatrix` analogous to our `IndexVec` (where I suppose it would have two type parameters instead of just one), so that we could use an index type here instead of `usize`... But, that need not  be part of this PR.

Reply:
I have done a big refactoring of this (in particular, encapsulating the `BitMatrix`) on nll-master, but I'm not inclined to bother rebasing it back earlier at this point.